refactoring.api/src/org/netbeans/modules/refactoring/api/RefactoringSession.java

```diff
@@ -246,27 +223,20 @@ private Problem reallyUndoRefactoring(boolean saveAfterDone) {
                     f.undoChange();
                 }
             }
-            for (Transaction commit : SPIAccessor.DEFAULT.getCommits(bag)) {
-                SPIAccessor.DEFAULT.check(commit, true);
-            }
-            UndoableWrapper wrapper = MimeLookup.getLookup("").lookup(UndoableWrapper.class);
+            SPIAccessor.DEFAULT.getCommits(bag).forEach(commit -> SPIAccessor.DEFAULT.check(commit, true));
+            UndoableWrapper undoableWrapper = MimeLookup.getLookup("").lookup(UndoableWrapper.class);
             for (ListIterator<Transaction> commitIterator = commits.listIterator(commits.size()); commitIterator.hasPrevious();) {
                 final Transaction commit = commitIterator.previous();
-                setWrappers(commit, wrapper);
+                undoableWrapper.setActive(true, this);
                 commit.rollback();
-                unsetWrappers(commit, wrapper);
-            }
-            wrapper.close();
-            for (Transaction commit : SPIAccessor.DEFAULT.getCommits(bag)) {
-                SPIAccessor.DEFAULT.sum(commit);
+                undoableWrapper.setActive(false, null);
             }
-
-            while (it.hasPrevious()) {
+            undoableWrapper.close();
+            SPIAccessor.DEFAULT.getCommits(bag).forEach((commit) -> SPIAccessor.DEFAULT.sum(commit));
+            while (internalListIterator.hasPrevious()) {
                 fireProgressListenerStep();
-                RefactoringElementImplementation element = (RefactoringElementImplementation) it.previous();
-                if (element.isEnabled() && !((element.getStatus() == RefactoringElement.GUARDED) || (element.getStatus() == RefactoringElement.READ_ONLY))) {
-                    element.undoChange();
-                }
+                RefactoringElementImplementation element = (RefactoringElementImplementation) internalListIterator.previous();
```

Comment:
The type cast to `RefactoringElementImplementation` could be removed.

Reply:
Well spotted. Not bad!